### PR TITLE
Content security policy

### PR DESCRIPTION
**Pull request by erikperkins**
Fixes issue #17.

Harden app against cross-site scripting using Content-Security-Policy header.

### CODE REVIEW

File: test/models/image_test.rb [hunk 1]
```diff
@@ -8,10 +8,6 @@ class ImageTest < ActiveSupport::TestCase
     assert_predicate image, :valid?
   end
 
-  test 'create invalid image' do
```
Comment (bboe) [inline test/models/image_test.rb]: I'm not sure why this deletion is in this branch. Maybe rebasing is necessary?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,4 +2,20 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+  before_filter :content_security_policy
```
Comment (bboe) [inline app/controllers/application_controller.rb]: While this approach absolutely works, is a bit more kludgy than necessary. We can set default headers instead:
http://edgeguides.rubyonrails.org/security.html#default-headers

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -2,4 +2,20 @@ class ApplicationController < ActionController::Base
...
+
+  private
+
+  def content_security_policy
+
+    policy = %q{default-src 'none'; script-src 'self'; img-src 'self'; style-src 'self';}
+
+    # A blog from 2013 suggested covering all major browsers this way
+    response.headers['Content-Security-Policy'] = policy
+    response.headers['X-Content-Security-Policy'] = policy
```
Comment (bboe) [inline app/controllers/application_controller.rb]: I didn't realize how relatively recently the `Content-Security-Policy` header became standard. Good catch!

I'm curious how the older browsers support/enforce the policy. If we're going to be restrictive explicitly for those older browsers, we probably want to test using them as well every time we make a change. For simplicity I think it's okay if we're not as strict when our users use an older browser.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,4 +2,20 @@ class ApplicationController < ActionController::Base
...
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+  before_filter :content_security_policy
+
+  private
+
+  def content_security_policy
+
+    policy = %q{default-src 'none'; script-src 'self'; img-src 'self'; style-src 'self';}
```
Comment (bboe) [inline app/controllers/application_controller.rb]: Did you test this policy against the site? Use the developer console in chrome to see if there are any errors that need to be resolved.

File: test/models/image_test.rb (hunk 1 above)
Comment (erikperkins) [inline test/models/image_test.rb]: Addressed by rebase

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (erikperkins) [inline app/controllers/application_controller.rb]: Addressed in f8fa93d6878c47b26257fd9780e4a40dd1f0e55a

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (erikperkins) [inline app/controllers/application_controller.rb]: Addressed in 508719763b288a4cc8922a18667c13497d30677c

**Comment by bboe** [comment]
Changes look good. As I mentioned before with the addition of the old CSP headers that it would be good to verify that adding the specific CSP doesn't break anything on those older browsers. This verification process _should_ be repeated for every future story as the introduction of this code makes those "supported" browsers.

As a result I think it might be better to only support the modern CSP header, as I don't think it's worth our time to verify every change. Whether or not you remove the other headers, I think the commits can be squashed at this point.

**Comment by erikperkins** [comment]
Fixes squashed.

How does one go about testing code on the hundreds of old versions of browsers? Only one is installed on my laptop :)

**Comment by bboe** [comment]
> How does one go about testing code on the hundreds of old versions of browsers? Only one is installed on my laptop :)

I would think that testing the most recent version of each browser before it supports the standard headers would be sufficient. Doing it with chrome is tricky as it autoupdates, but it should be simpler with firefox. IE requires a virtual machine. It's not worth the effort in my opinion. The acceptance tests said nothing about supporting non-modern browsers.

**Comment by erikperkins** [comment]
Okay, I'll take it that it's okay to merge into master now, unless you say otherwise. I'll leave the other browser-specific headers in.

**Pull request merged**